components/eamxx/src/physics/mam/eamxx_mam_aci_functions.hpp

```diff
@@ -0,0 +1,726 @@
+#ifndef EAMXX_MAM_ACI_FUNCTION_HPP
+#define EAMXX_MAM_ACI_FUNCTION_HPP
+
+#include <ekat/kokkos/ekat_subview_utils.hpp>
+#include <mam4xx/mam4.hpp>
+#include <share/util/scream_common_physics_functions.hpp>
+
+namespace scream {
+
+namespace {
+
+KOKKOS_INLINE_FUNCTION
+void compute_w0_and_rho(const haero::ThreadTeam &team,
+                        const MAMAci::const_view_2d omega,
+                        const MAMAci::const_view_2d T_mid,
+                        const MAMAci::const_view_2d p_mid, const int icol,
+                        const int top_lev, const int nlev,
+                        // output
+                        MAMAci::view_2d w0, MAMAci::view_2d rho) {
+  // Get physical constants
+  using C                      = physics::Constants<Real>;
+  static constexpr auto gravit = C::gravit;  // Gravity [m/s2]
+  // Gas constant for dry air [J/(kg*K) or J/Kg/K]
+  static constexpr auto rair = C::Rair;
+  Kokkos::parallel_for(
+      Kokkos::TeamThreadRange(team, 0u, top_lev), KOKKOS_LAMBDA(int kk) {
+        w0(icol, kk)  = 0;
+        rho(icol, kk) = -999.0;
+      });
```

Comment:
I'll own up to this one.  I thought that since the loop indexes do not overlap, two loops looked better than one loop with an if statement.  Just a style choice.

Reply:
I guess avoiding the if-check is the main driver here (@odiazib and @overfelt can comment more about it). Although, it may not make a dramatic change in performance, avoiding if-check should help these kernels to run faster.

Reply:
Yes, the 0u is assumed.  

The use of KOKKOS_LAMBDA evaluates to [=] __host__ __device__ but since we are already in device code, the Cuda decorations might be redundant and [&] better if there is no warning about calling _host_ from _device_? I'm not sure.

Reply:
I changed the inner kernels as I got the following error on GPU tests:
```
An extended __host__ __device__ lambda cannot be defined inside an extended __host__ __device__  lambda expression("operator()").
```
I googled and google suggested to use `[=]` instead of `KOKKOS_LAMBDA`

Reply:
Since you are already on device, you should use `[&]` instead of `[=]`. The latter is _needed_ for the outermost lambda (you can't take references of host objects). But once you are on device, all the visible stuff is already on device, so might as well use references rather than copies.